### DISCUSSION TITLE
midtrain stuck

### DISCUSSION THREAD

**Discussion by couldn** [karpathy/nanochat#200] [General]
I'm stuck at midtrain, and the logs have stopped printing. I'm using 8×A100 GPUs, and it has been 12 hours already. Is this normal?

2025-10-16 19:59:50,381 - nanochat.checkpoint_manager - INFO - No model tag provided, guessing model tag: d20
2025-10-16 19:59:50,381 - nanochat.checkpoint_manager - INFO - Loading model from /root/.cache/nanochat/base_checkpoints/d20 with step 21400
2025-10-16 19:59:51,395 - nanochat.checkpoint_manager - INFO - Building model with config: {'sequence_len': 2048, 'vocab_size': 65536, 'n_layer': 20, 'n_head': 10, 'n_kv_head': 10, 'n_embd': 1280}
Tokens / micro-batch / rank: 32 x 2048 = 65,536
Tokens / micro-batch: 524,288
Total batch size 524,288 => gradient accumulation steps: 1
Scaling the LR for the AdamW parameters ∝1/√(1280/768) = 0.774597
Muon: Grouping 80 params of shape torch.Size([1280, 1280]), device cuda:0, dtype torch.float32
Muon: Grouping 20 params of shape torch.Size([1280, 5120]), device cuda:0, dtype torch.float32
Muon: Grouping 20 params of shape torch.Size([5120, 1280]), device cuda:0, dtype torch.float32

**Comment by couldn** [comment]
stuck in here

train_dataset = TaskMixture([
    SmolTalk(split="train"), # 460K rows of general conversations
    MMLU(subset="auxiliary_train", split="train"), # 100K rows of multiple choice problems drawn from ARC, MC_TEST, OBQA, RACE
    GSM8K(subset="main", split="train"), # 8K rows teaching simple math and (calculator) tool use
])

**Comment by Milpo1** [comment]
@couldn Im having a similar issue. Have you found a solution?

**Comment by samerGMTM22** [comment]
It seems mybttaining is stilucknin mid training wanb is showing runs but no logs will give time and see before shutting GPUs as it is expensive... Any solution please

**Reply by samerGMTM22** [reply]
I solved it for my self, as I swapped mid training data sets that created and issue in sharding size

Problem:
The pile-of-law/pile-of-law dataset (I am using for my model) preparation script created too many small shards, likely due to:
-Row group alignment forcing shard boundaries early
-Each shard hitting 250M characters but compressing down to 65MB
-Total dataset: 7.60 GB (30B characters) ÷ 120 shards = 63MB/shard

Final Solution: Shard Consolidation
Implementation
Created consolidation script that merges 5 old shards → 1 new shard:
python# Consolidation parameters
SHARDS_PER_NEW = 5  # Merge 5 shards into 1
TARGET_SHARDS = 24  # 120 / 5 = 24

# Results:
24 shards averaging 325 MB each
Total size: 7.8 GB (same data, better structure)
Row group size: 1024 (maintained)
Compression: zstd level 3 (maintained)
```

### Consolidation Results
```
Created shard_00000.parquet: 25600 docs, 326.9 MB
Created shard_00001.parquet: 24576 docs, 321.4 MB
...
Created shard_00023.parquet: 25600 docs, 323.1 MB

✓ 24 consolidated shards
✓ ~325 MB average size (5x larger than original)
✓ 80% reduction in file count
Performance Improvement
Before (120 shards):

First batch: >3 minutes (timeout)
8-GPU training: FAILED

After (24 shards):

First batch: <30 seconds ✓
Step 1: 19.7s, Step 2: 1.5s ✓
MFU: 53.67% ✓
8-GPU training: SUCCESS ✓

So, with this I validated it was the problem and ran complete mid training which now is progressing really well and will complete soon.

Honestly, one hickup that neither Codex nor Claude code figured; but it took me to think of this as a potential impact then claude suggested the fix
